### ISSUE TITLE
weighted use_backends based on internal stats

### ISSUE THREAD

**Issue by bedis** [haproxy/haproxy#475]
Currently, in HAProxy, the `use_backend` directive points to a single backend, which is good for "routing".
That said, more and more often, I see the need for balancing the backends and so the current `use_backend` becomes limited.
Of course, we can use stats, math and some use_backend rules coupled to some ACLs:

> frontend f
> bind 127.0.0.1:9000
> http-request set-var(txn.b1Rate) be_sess_rate(b1),debug
> http-request set-var(txn.b2Rate) be_sess_rate(b2),debug
> http-request set-var(txn.Rate)   var(txn.b1Rate),add(txn.b2Rate),debug
> http-request set-var(txn.ratio)  int(100),div(30)                       # ratio in %
> use_backend b1 if { var(txn.Rate),div(txn.ratio),sub(txn.b1Rate) -m int gt 0 } or { var(txn.ratio) eq 1 }  # might be a smarter way to make the computation
> use_backend b2
>
> backend b1
>  server s1 127.0.0.1:9001 check
>
>backend b2
> server s2 127.0.0.1:9001 check

But I think it would be much easier to have this type of thing:

> use_backends b1:30 b2:100  be_sess_rate

syntax could be:
`use_baceknds <backend name>:<weight or percentage> <backend name>:<weight or percentage> <be metric observed>`

The idea is to allow using any metrics exposed by HAProxy like sess rate, cur_conn, etc...

Cherry on the cake if we could monitor the error rate or queues and give options to the admin to influence the weighing based on these metrics too.

**Comment by wtarreau** [comment]
No, I strongly disagree here. You're suggesting to roll back to haproxy 1.0 which had no health checks, and from there we'll re-create all the painful things we've gone through over 19 years. Because implementing load balancing, whatever form it disguises into, **does requires health checking**. This means that the **backend's health** will be needed. And a backend's health is not a boolean but a metric (total weight or # servers or active vs backup). This also means having to implement a second stage of load balancing with all its horde of algorithms (with hashes being the ones which will be requested the most). Then users will complain that only the first request goes to the right server and we'll have to factor in cookie stickiness there. Then we'll need to deal with stick tables in the frontends to select the appropriate backend.

In fact what you're doing is **exactly** stacking two layers of haproxy except that you **currently** don't care about the health of the second layer nor the decisions that made you enter it, which makes you think you can thus completely force this into the same layer. But that's a wrong assumption in my opinion, better stick to two real layers for this.

For the occasional cases where you're certain you don't care about the farm's health, you can already achive what you need using a dynamic rule and your operations above. You may even call some Lua to do the computation for you and you have all the flexibility needed (including taking into accound the backend's state) like this:
```
    frontend f
        use_backend %[lua.bestbk(b1,b2)]
```
and let the Lua function perform the operation for you based on whatever criteria are needed, including considering the number of servers/total weight/number of active connections/backend servers' average response time etc. With your form above, you actually hard-code a specific configuration into the code, that cannot be modified anymore for all other use cases, and making it adaptable will actually require far more configuration to work around all default behaviours decided in february 2020.

Now with all this said, you might or might not have an interesting use case that could possibly be improved differently. Load balancing between backends makes no sense by definition in the general case, but backends are a way to have multiple server groups for the same application, and it's exactly in this case that one might want to use multiple backends for the same application (multiple datacenters). Then the question should be "why would we load balance on a given criterion ?". Couldn't this decision be implemented into the existing backends at the server level by simply modulating the servers' weights ? Maybe it would even be smarter. The classical request of two backends because of two DCs is already addressed by setting per-server weights. When instead the purpose is to use it with the old and the new version of an application being progressively migrated, it **should** work within a single backend or it's a massive application change since a part of the logic differs and requires incompatible rules in a different backend. When doing this with incompatible configs, it's exactly where two explicit layers should be set, because there's always something that will not properly map to a change like above (especially the new application crashing or queuing excessively without allowing to automatically throttle).

Just my two cents.

**Comment by bedis** [comment]
Sorry to wake up this issue :)
The use case was for cannary deployments, so basically ramping up a new version of the same application. This is supposed to also embed kind of "monitoring" of new version and stop sending traffic to too many (threshold to be defined) errors happen
